Added support for embedding indexed images

RedShy opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Some guidelines are provided like this, in HTML comments, to expedite the code review before merging your contribution.

First, please make sure that you have read the documentation page on fpdf2 development:
https://pyfpdf.github.io/fpdf2/Development.html

If you're new to contributing to open source projects,
you might find this free video course helpful: http://kcd.im/pull-request
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

I added the support for embedding indexed images in a PDF. The code was already there, I just had to do some minor fixes. I also added a test to cover the new feature.

There is a problem though: **previous tests now breaks.**

The tests processed indexed images by converting them to RGBA and then embedding them in the PDF, now instead they embeds them directly as indexed images.

Not sure how to proceed now.



<!-- Have you done all of these things?  -->
**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [ ] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [ ] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

Lucas-C commented on test/image/png_indexed/test_png_indexed.py:
```diff
@@ -0,0 +1,46 @@
+    pdf.add_page()
+
+    # palette images (P)
+    pdf.image(HERE / "palette_1.png", x=10, y=10, w=50, h=50, link=None)
```
The `link=None` are not needed

Lucas-C reviewed (approved):
Good job really! Well done, the code is clean and you seemingly reduced the size of the PDFs generated when they embed indexed images.

>  My idea is to generate the PDF again using generate=True.

Seems like the best approach here.
Just make sure the new reference PDF files are smaller than the previous ones.

RedShy commented:
I observed that if I change the image filter to `"JPXDecode"` or `"DCTDecode"` it throws an error when executing `img.save(compressed_bytes, format="JPEG2000")` so I forced the conversion to RGBA in case of image filter != "FlateDecode" and image mode is "P" or "PA".

It also seems that Pillow doesn't work well with "PA" images.
I used pngquant to create an indexed-alpha PNG, but Pillow still opens it as a "P" image without alpha channel. I then tried to convert it to "PA" but the result embedded image is:
![](https://user-images.githubusercontent.com/24901693/169567939-c92146ff-0f5b-4569-8148-45b5322711b0.png)

To make sure that the code regarding the alpha channel works correctly, I manually set the alpha channel:
```python
palette = img.palette
a = np.asarray(img)
for r in range(a.shape[0]):
   for c in range(a.shape[1]):
       a[r][c][1] = 100
img = Image.fromarray(a, mode="PA")
img.palette = palette
```
And the result image is without artifacts:
![](https://user-images.githubusercontent.com/24901693/169568728-b1f003b1-213c-43a1-9522-7bea00b2c953.png)

Lucas-C commented:
Thank you for you work on this!
I think I'll take the time to review this on Monday morning.
Have a good week-end 😊

RedShy commented:
> I think I'll take the time to review this on Monday morning.

Sure! Thank you! Have a good week-end you too 😊

Lucas-C commented on test/errors/test_FPDF_errors.py:
```diff
@@ -105,3 +106,15 @@ def test_repeated_calls_to_output(tmp_path):
+    class Object:
+        pass
+
+    img = Object()
```
Why building a mock image here, and not simply passing a real `PIL.Image`?

RedShy replied:
I tried to write minimal code to raise the exception. But yes it's better to test other things at the same time

Lucas-C commented on test/image/image_info.json:
```diff
@@ -1,1612 +1 @@
-    "dp": "/Predictor 15 /Colors 3 /BitsPerComponent 8 /Columns 32"
-  }
-}
+{"basi0g01.png": {"w": 32, "h": 32, "cs": "DeviceRGB", "bpc": 8, "f": "FlateDecode", "dp": "/Predictor 15 /Colors 3 /BitsPerComponent 8 /Columns 32", "trns": ""}, "basi0g02.png": {"w": 32, "h": 32, "cs": "DeviceGray", "bpc": 8, "f": "FlateDecode", "dp": "/Predictor 15 /Colors 1 /BitsPerComponent 8 /Columns 32", "trns": ""}, "basi0g04.png": {"w": 32, "h": 32, "cs": "DeviceGray", "bpc": 8, "f": "Fl ...
```
Could you please preserve the original JSON formatting?
Currently the diff does not make it possible to understand the changes you made...

RedShy replied:
Yes you are right, I didn't thought about the formatting

Lucas-C commented on test/image/png_images/test_png_file.py:
```diff
@@ -23,6 +23,6 @@ def test_insert_png_files(tmp_path):
     for path in sorted(HERE.glob("*.png")):
         if path.name not in not_supported:
             pdf.add_page()
-            pdf.image(str(path), x=0, y=0, w=0, h=0, link=None)
```
👍

Lucas-C commented on test/image/test_image_info.py:
```diff
@@ -29,25 +29,41 @@ def break_down_filename(image):
+            short_info = {}
+            for k, v in info.items():
+                if k in short_keys:
+                    short_info[k] = v.decode("latin-1") if isinstance(v, bytes) else v
```
Why is this call to `.decode()` needed?

RedShy replied:
Because now `pal` is a sequence of bytes and when I tried to serialize it in JSON it threw `TypeError: Object of type 'bytes' is not JSON serializable` so before serialize it, I called `.decode()`

I also added `smask` in the `short_info`, it was not necessary, but I thought that could be tested too

Lucas-C replied:
OK, alright!

RedShy commented on test/image/png_indexed/test_png_indexed.py:
```diff
@@ -0,0 +1,36 @@
+    # palette with alpha images (PA)
+    pdf.set_image_filter("FlateDecode")
+    pdf.image(
+        Image.open(HERE / "palette_alpha_1.png").convert("PA"), x=10, y=80, w=50, h=50
```
I agree with you. Don't know how to act here, seems to be a problem with the conversion in "PA". 
Maybe it's better to drop the support for "PA" images? When I open an indexed-alpha image, Pillow opens it in "P" mode.

Lucas-C replied:
> Maybe it's better to drop the support for "PA" images?

Yes that is probably best, or maybe check Pillow source code to find PA reference images that the lib actually supports...

RedShy commented:
> It may be worth reporting them this issue btw.

In the next days I will try open an issue in the Pillow's github page

> I think you did well and this PR!

Thank you! 

> What was the error exactly?

`image_filter=JPXDecode`: `OSError: encoder error -2 when writing image file` executing `img.save(compressed_bytes, format="JPEG2000")` 
```
Traceback (most recent call last):
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\ImageFile.py", line 500, in _save
    fh = fp.fileno()
io.UnsupportedOperation: fileno

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "D:\Progetti OpenSource\playground_fpdf2\coverage.py", line 80, in <module>
    test_png_indexed_files(None)
  File "D:\Progetti OpenSource\playground_fpdf2\coverage.py", line 53, in test_png_indexed_files
    pdf.image(HERE / "palette_2.png", x=80, y=10, w=50, h=50)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\fpdf.py", line 299, in wrapper
    return fn(self, *args, **kwargs)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\fpdf.py", line 3235, in image
    info = get_img_info(img, self.image_filter)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\image_parsing.py", line 92, in get_img_info
    info["data"] = _to_data(img, image_filter)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\image_parsing.py", line 151, in _to_data
    img.save(compressed_bytes, format="JPEG2000")
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\Image.py", line 2300, in save
    save_handler(self, fp, filename)
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\Jpeg2KImagePlugin.py", line 348, in _save
    ImageFile._save(im, fp, [("jpeg2k", (0, 0) + im.size, 0, kind)])
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\ImageFile.py", line 526, in _save
    raise OSError(f"encoder error {s} when writing image file") from exc
OSError: encoder error -2 when writing image file
```

`image_filter=DCTDecode`: `OSError: cannot write mode P as JPEG` executing `img.save(compressed_bytes, format="JPEG2000")`
```
Traceback (most recent call last):
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\JpegImagePlugin.py", line 633, in _save
    rawmode = RAWMODE[im.mode]
KeyError: 'P'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "D:\Progetti OpenSource\playground_fpdf2\coverage.py", line 80, in <module>
    test_png_indexed_files(None)
  File "D:\Progetti OpenSource\playground_fpdf2\coverage.py", line 56, in test_png_indexed_files
    pdf.image(HERE / "palette_3.png", x=150, y=10, w=50, h=50)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\fpdf.py", line 299, in wrapper
    return fn(self, *args, **kwargs)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\fpdf.py", line 3235, in image
    info = get_img_info(img, self.image_filter)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\image_parsing.py", line 92, in get_img_info
    info["data"] = _to_data(img, image_filter)
  File "D:\Progetti OpenSource\forked_fpdf2\fpdf2\fpdf\image_parsing.py", line 146, in _to_data
    img.save(compressed_bytes, format="JPEG")
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\Image.py", line 2300, in save
    save_handler(self, fp, filename)
  File "C:\Program Files\Python3.10\lib\site-packages\PIL\JpegImagePlugin.py", line 635, in _save
    raise OSError(f"cannot write mode {im.mode} as JPEG") from e
OSError: cannot write mode P as JPEG
```

RedShy commented:
In the last days I searched in the code of Pillow and I discovered I had the **false assumption** that "P" images don't support transparency and "PA" yes. Instead, both support transparency:
- "P" support transparency through the attribute `info["transparency"]`: ([source](https://github.com/python-pillow/Pillow/blob/26c79343ec864d4b326100c543c4a1eb21c5adb5/src/PIL/PngImagePlugin.py#L474))
	- **int:** is the color in the palette regarded as "transparent", others are full opaque
	- **bytes**: a byte string with one alpha value for each palette entry
- "PA" support transparency through one alpha value for each pixel

Seems that Pillow has a problem in generating the alpha layer when converting from "P" to "PA". This doesn't happen in the conversion "P" to "RGBA". I opened the [issue](https://github.com/python-pillow/Pillow/issues/6336) and the relative [PR](https://github.com/python-pillow/Pillow/pull/6337) that should fix it.


Regarding fpdf2 and "P" images, currently the transparency is not ported in the PDF and for doing this I see 2 approaches:
- Implement a homebrew solution using `info["transparency"`]
- Use Pillow to get the alpha channel of "P" images and reuse the existing code. This can be achieved by converting the image in "RGBA" and extracting the alpha channel. This is the solution I implemented now. 
		
Regarding "PA" images, the code in the fpdf2 seems to work correctly if the alpha layer is correctly set. So I suggest we keep supporting them.
I added tests in which I test the "PA" images by inserting the alpha channel extracted from the converted "RGBA" image and they are displayed correctly in the PDF.

(I also changed the sample images in the test putting some pics of flowers that I took near my house 😊)

Lucas-C commented:
Wow, you did an excellent job!
You even fixed the bug upstream in Pillow, awesome!

I'm going to fix the minor file conflicts that prevent this PR from being merged,
and see if it can be closed afterwards.

Lucas-C commented on fpdf/image_parsing.py:
```diff
@@ -79,6 +87,30 @@ def get_img_info(img, image_filter="AUTO", dims=None):
+        info["data"] = _to_data(img, image_filter)
+        info["pal"] = img.palette.palette
+
+        # che if the P image has transparency
```
Minor typo: `che` -> `check`

Lucas-C commented on test/errors/test_FPDF_errors.py:
```diff
@@ -107,6 +109,14 @@ def test_repeated_calls_to_output(tmp_path):
+    img = Image.open(HERE / "flowers.png")
+    image_filter = "N/A"
+    with pytest.raises(FPDFException) as error:
+        _to_data(img, image_filter)
```
`_to_data` is currently a module-private function: could we instead test the parent `get_img_info()` public function?

RedShy replied:
Yes right! I will keep in mind next times: better to test public functions instead of  module-private functions

Lucas-C commented on test/image/png_indexed/test_png_indexed.py:
```diff
@@ -0,0 +1,103 @@
+        # open the image as RGBA
+        img_RGBA = Image.open(path_png).convert("RGBA")
+
+        # estract the alpha channel
```
Minor typo: `estract` -> `extract`

Lucas-C commented on test/image/png_indexed/test_png_indexed.py:
```diff
@@ -0,0 +1,103 @@
+
+    # PA images
+    img = Image.open(HERE / "flower2.png").convert("PA")
+    img = insert_alpha_channel_from_RGBA(img, HERE / "flower2.png")
```
Well done crafting this solution with `insert_alpha_channel_from_RGBA`.

Just a though: couldn't we do this only once, then call `img.save('flower2-pa.png')`
and then have this test uses `flower2-pa.png`?

This would reduce the test length and make it a bit easier to grasp.

RedShy replied:
I agree, I tried to do that after reading your comment, but I get `OSError: cannot write mode PA as PNG` when calling `img.save('flower2-pa.png')`

Lucas-C replied:
Alright, you can keept the code as it is then 😊

Lucas-C commented on test/image/png_indexed/test_png_indexed.py:
```diff
@@ -0,0 +1,103 @@
+    assert Image.open(HERE / "flower1.png").mode == "P", "img.mode is not P"
+
+    pdf.set_image_filter("FlateDecode")
+    pdf.image(HERE / "flower1.png", x=10, y=10, w=50, h=50)
```
You have really beautiful flowers in your garden ^^
Thanks for sharing!

RedShy replied:
Thank you!

Lucas-C commented on test/errors/test_FPDF_errors.py:
```diff
@@ -107,6 +109,14 @@ def test_repeated_calls_to_output(tmp_path):
+        _to_data(img, image_filter)
+    assert str(error.value) == f'Unsupported image filter: "{image_filter}"'
+
+    
```
Seems like Pylint does not like some trailing whitespaces here:
```
test/errors/test_FPDF_errors.py:119:0: C0303: Trailing whitespace (trailing-whitespace)
```

RedShy replied:
I added the pre-commit hook with pylint and black for next times

RedShy commented:
> Wow, you did an excellent job!
> You even fixed the bug upstream in Pillow, awesome!

Thank you, that means a lot to me! it's really rewarding and encourages me to do more! 😊

Lucas-C commented:
It seems like some tests are failing:
```
FAILED test/html/test_html.py::test_img_inside_html_table - AssertionError: c...
FAILED test/html/test_html.py::test_img_inside_html_table_without_explicit_dimensions
FAILED test/html/test_html.py::test_img_inside_html_table_centered - Assertio...
FAILED test/html/test_html.py::test_img_inside_html_table_centered_with_align
FAILED test/image/test_image_info.py::test_get_img_info - assert {'bpc': 8,\n...
FAILED test/image/png_images/test_png_file.py::test_insert_png_files - Assert...
================= 6 failed, 914 passed, 18 warnings in 42.17s ==================
```

RedShy commented:
> Do you have the same errors on your computer?

Yes I forgot to run **all** the tests again

Lucas-C commented:
Merged! Thank you @RedShy 😊

Lucas-C commented:
This feature has been included in the new [v2.5.5](https://github.com/PyFPDF/fpdf2/releases/tag/2.5.5) release!
